pkg/cmd/dev/util.go

```diff
@@ -105,11 +107,74 @@ func addCommonTestFlags(cmd *cobra.Command) {
 	cmd.Flags().Duration(timeoutFlag, 0*time.Minute, "timeout for test")
 }
 
-func ensureBinaryInPath(bin string) error {
+func ensureBinaryInPath(dev *dev, bin string) error {
 	if !isTesting {
-		if _, err := exec.LookPath(bin); err != nil {
+		if _, err := dev.exec.LookPath(bin); err != nil {
 			return errors.Newf("Could not find %s in PATH", bin)
 		}
 	}
 	return nil
 }
+
+// setupPath removes the ccache directory from PATH to prevent it writing files
+// outside of the bazel sandbox. This function is called only once to prevent
+// multiple unnecessary calls.
+func setupPath(dev *dev) error {
+	var once sync.Once
+	var err error
+	once.Do(func() {
+		err = setupPathReal(dev)
+	})
+	return err
+}
+
+// setupPathReal uses a list of known compiler names to check if they are
+// symlinks to `ccache`.
+func setupPathReal(dev *dev) error {
+	knownCompilers := []string{"cc", "gcc", "c++", "g++", "clang", "clang++"}
+	// datadriven uses Fprintln() to construct commands, adding '\n' to the end.
+	// Trim the output here and in other places that don't expect the extra new line
+	// at the end.
+	origPath := strings.TrimSuffix(dev.os.Getenv("PATH"), "\n")
+	pathEntries := strings.Split(origPath, string(os.PathListSeparator))
+	for i, entry := range pathEntries {
+		pathEntries[i] = filepath.Clean(entry)
+	}
+	for _, compiler := range knownCompilers {
+		compilerPath, err := dev.exec.LookPath(compiler)
+		if err != nil {
+			continue
+		}
+		compilerPath = strings.TrimSuffix(compilerPath, "\n")
+		compilerDir, _ := filepath.Split(compilerPath)
+		compilerDir = filepath.Clean(compilerDir)
+		compilerResolvedPath, err := dev.os.Readlink(compilerPath)
+		if err != nil {
+			// Skip broken symlinks and real binaries
+			continue
+		}
+		compilerResolvedPath = strings.TrimSuffix(compilerResolvedPath, "\n")
+		_, file := filepath.Split(filepath.Clean(compilerResolvedPath))
+		if file != "ccache" {
+			continue
+		}
+		// The compiler points to ccache, remove it from PATH
+		var newPathEntries []string
+		for _, dir := range pathEntries {
+			if compilerDir != dir {
+				newPathEntries = append(newPathEntries, dir)
+			}
+		}
+		newPath := strings.Join(newPathEntries, string(os.PathListSeparator))
+		if origPath == newPath {
+			log.Printf("WARNING: PATH did not change: %s", origPath)
+		}
+		log.Printf("INFO: removed ccache directory from PATH. New PATH: %s", newPath)
```

Comment:
Nit: This should be silent under most circumstances. Printing this out every time someone invokes `dev` is not helpful. If you still want the message you can do `if debugVar { log.Printf(...) }`

Reply:
Removed.